weld/ast.rs

```diff
@@ -75,6 +75,181 @@ impl fmt::Display for ScalarKind {
 }
 
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum BuilderImplementationKind {
+    Local,
+    Global
+}
+
+impl fmt::Display for BuilderImplementationKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ast::BuilderImplementationKind::*;
+        let text = match *self {
+            Local => "local",
+            Global => "global",
+        };
+        f.write_str(text)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Annotations {
+    builder_implementation: Option<BuilderImplementationKind>,
+
+    predicate: Option<bool>,
+    vectorize: Option<bool>,
+    tile_size: Option<i32>,
+    grain_size: Option<i32>,
+
+    size: Option<i32>,
+    branch_selectivity: Option<i32>, // Fractions of 10,000.
+    num_keys: Option<i32>,
+}
+
+impl Annotations {
+    pub fn new() -> Annotations {
+        return Annotations {
+            builder_implementation: None,
+            predicate: None,
+            vectorize: None,
+            tile_size: None,
+            grain_size: None,
+            size: None,
+            branch_selectivity: None,
+            num_keys: None,
+        }
+    }
+
+    pub fn get_builder_implementation(&self) -> Option<BuilderImplementationKind> {
+        self.builder_implementation.clone()
+    }
+
+    pub fn set_builder_implementation(&mut self,
+                                      builder_implementation: BuilderImplementationKind) {
+        self.builder_implementation = Some(builder_implementation);
+    }
+
+    pub fn get_predicate(&self) -> Option<bool> {
+        self.predicate.clone()
+    }
+
+    pub fn set_predicate(&mut self, predicate: bool) {
+        self.predicate = Some(predicate)
+    }
+
+    pub fn get_vectorize(&self) -> Option<bool> {
+        self.vectorize.clone()
+    }
+
+    pub fn set_vectorize(&mut self, vectorize: bool) {
+        self.vectorize = Some(vectorize)
+    }
+
+    pub fn get_tile_size(&self) -> Option<i32> {
+        self.tile_size.clone()
+    }
+
+    pub fn set_tile_size(&mut self, tile_size: i32) {
+        self.tile_size = Some(tile_size)
+    }
+
+    pub fn get_grain_size(&self) -> Option<i32> {
+        self.grain_size.clone()
+    }
+
+    pub fn set_grain_size(&mut self, grain_size: i32) {
+        self.grain_size = Some(grain_size)
+    }
+
+    pub fn get_size(&self) -> Option<i32> {
+        self.size.clone()
+    }
+
+    pub fn set_size(&mut self, size: i32) {
+        self.size = Some(size)
+    }
+
+    pub fn get_selectivity(&self) -> Option<i32> {
+        self.branch_selectivity.clone()
+    }
+
+    pub fn set_selectivity(&mut self, selectivity: i32) {
+        self.branch_selectivity = Some(selectivity)
+    }
+
+    pub fn get_num_keys(&self) -> Option<i32> {
+        self.num_keys.clone()
+    }
+
+    pub fn set_num_keys(&mut self, num_keys: i32) {
+        self.num_keys = Some(num_keys)
+    }
+
+    pub fn is_empty(&self) -> bool {
```

Comment:
Is there a way to do this so the compiler throws an error if we add new annotations and forget to add them here (e.g., something like a `MatchError`?) Maybe it's not possible.